tests/test_connections/test_ledger/test_ledger.py

```diff
@@ -69,30 +84,193 @@ async def test_not_hanging(self) -> None:
+        )
+        cls.multiplexer.connect()
+        # the ledger connection's connect() is called by the multiplexer
+        # once a connection is ready, `receive()` is called by the multiplexer
+        cls.ledger_connection = cast(LedgerConnection, cls.multiplexer.connections[0])
+        cls.ledger_api_dialogues = LedgerApiDialogues(SOME_SKILL_ID)
+
+    @classmethod
+    def teardown(cls) -> None:
+        """Tear down the multiplexer."""
+        if cls.multiplexer.is_connected:
+            cls.multiplexer.disconnect()
+        cls.running_loop.call_soon_threadsafe(cls.running_loop.stop)
+        cls.thread_loop.join()
+
+    def create_ledger_dialogues(self) -> Tuple[LedgerApiMessage, LedgerApiDialogues]:
+        """Create a dialogue."""
+        return cast(
+            Tuple[LedgerApiMessage, LedgerApiDialogues],
+            self.ledger_api_dialogues.create(
+                counterparty=str(self.ledger_connection.connection_id),
+                performative=LedgerApiMessage.Performative.GET_STATE,  # type: ignore
+                ledger_id=EthereumCrypto.identifier,
+                address=EthereumCrypto(ETHEREUM_KEY_DEPLOYER).address,
+                callable="get_block",
+                args=("latest",),
+                kwargs=Kwargs({}),
+            ),
+        )
+
+    @staticmethod
+    def create_envelope(request: LedgerApiMessage) -> Envelope:
+        """Create a dummy envelope."""
+        return Envelope(
+            to=request.to,
+            sender=request.sender,
+            message=request,
+        )
+
+    @pytest.mark.asyncio
+    async def test_not_hanging_with_multiplexer(self) -> None:
+        """Test that the connection does not hang and that the tasks cannot get blocked, using the multiplexer."""
+        # create configurations for the test, re bocking and non-blocking tasks' waiting times
+        blocking_time = 100
+        wait_time_among_tasks = 0.1
+        non_blocking_time = 1
+        tolerance = 1
+        assert (
+            non_blocking_time + tolerance + wait_time_among_tasks < blocking_time
+        ), "`non_blocking_time + tolerance + wait_time_among_tasks` should be less than the `blocking_time`."
+
+        # create a blocking task lasting `blocking_time` secs
+        request, _ = self.create_ledger_dialogues()
+        blocking_dummy_envelope = TestLedgerConnectionWithMultiplexer.create_envelope(
+            request
+        )
+        blocking_task = dummy_task_wrapper(
+            blocking_time,
+            LedgerApiMessage(
+                ledger_id=EthereumCrypto.identifier,
+                performative=LedgerApiMessage.Performative.ERROR,  # type: ignore
+                code=1,
+                message="",
+                data=b"blocking_task",
+            ),
+        )
+        with mock.patch.object(
+            self.ledger_connection, "_schedule_request", return_value=blocking_task
+        ):
+            self.multiplexer.put(blocking_dummy_envelope)
```

Comment:
The mocking does not work after f5a32ce2619122c68977f6b8ca74227122b7c0ed and therefore the test fails on the first assertion.
This could be because of the multiplexer running the connection's `send` on a separate thread.

Reply:
Yes it does, that's why I think you should use the approach of a custom RequestDispatcher as mentioned before.